### DISCUSSION TITLE
Can ImageMagick fill in missing areas by smart scaling?

### DISCUSSION THREAD

**Discussion by octos** [ImageMagick/ImageMagick#2484] [Help] [answered]
Given a stitched panorama missing sky areas in the corner, is it possible for ImageMagick to fill those areas seamlessly by stretching edges of the sky?

![Panorama missing sky areas](https://archive.vn/IqSlW/037e6c8b5045ac1c352b4f4c00b72267840eb8c1.png)

**Comment by snibgo** [comment]
I call this "filling holes". See my pages [Filling holes](http://im.snibgo.com/fillholes.htm) and [Filling holes in priority order](http://im.snibgo.com/fillholespri.htm). This is also called "inpainting".

**Reply by octos** [reply]
Will these work on macOS, or do I need Windows with Cygwin?

**Comment by snibgo** [comment, marked as answer]
The process modules need IM to be rebuilt. For that, you need a C compiler. I use the GNU toolset from Cygwin, but they should work on any platform.

The scripts and sample commands are Windows BAT. Bash syntax is different.

**Comment by fmw42** [comment]
Here is one way, though it leaves some artifacts due to the square edges to the sky boundary.

line1: read the input and make the sky transparent and save in memory
line2: select the alpha channel and process to get the outline of the blue sky on transparent background
line3: scale the outline down to one line of blue and then scale up to the original image size as a blue image
line4: load the saved image with transparent sky and erode the alpha channel a little
line5: composite the two together and save the result

```
convert sky.png -alpha off -transparent "rgb(246,245,246)" -write mpr:img \
-channel a -negate -morphology dilate square:3 -morphology edgeout square:1 +channel \
-scale x1! -alpha off -scale 1194x538! -alpha off \
\( mpr:img -channel a -morphology erode square:3 +channel \) \
-compose over -composite result.png
```


![result](https://user-images.githubusercontent.com/13458218/91749997-30c3df00-eb77-11ea-80f1-65659d23d6ca.png)

**Reply by fmw42** [reply]
Here is an alternate version that smooths the sky image before compositing. However, since the right side of the sky image is black where the sky goes to the top of the input image, we have to fill the black with some measured sky color so that the blur does not blur the transition to black in the sky image. So I compute the average blue sky colore  and fill the black with that color before blurring. Getting the average sky color is done similar to how I process the image. I also just use pure white for the whitish area above the sky that is to be filled by including a fuzz value.

```
skycolor=`convert sky.png -alpha off -fuzz 5% -transparent white \
-channel a -negate -morphology dilate square:3 -morphology edgeout square:1 +channel \
-scale 1x1! -alpha off -format "%[pixel:u.p{0,0}]" info:`

convert sky.png -alpha off -fuzz 5% -transparent white -write mpr:img \
-channel a -negate -morphology dilate square:3 -morphology edgeout square:1 +channel \
-scale x1! -alpha off -scale 1194x538! -fill "$skycolor" -opaque black -blur 0x101 \
\( mpr:img -channel a -morphology erode square:3 +channel \) \
-compose over -composite result2.png

```

![result2](https://user-images.githubusercontent.com/13458218/91754810-e5153380-eb7e-11ea-8d01-e6c439040491.png)
